lr_face/data_providers.py

```diff
@@ -0,0 +1,146 @@
+import csv
+from dataclasses import dataclass
+from typing import Tuple
+import os
+
+import pandas as pd
+import cv2
+import numpy as np
+from sklearn.model_selection import GroupShuffleSplit
+
+
+@dataclass
+class Images:
+    y_train: np.ndarray
+    X_train: np.ndarray
+    y_calibrate: np.ndarray
+    X_calibrate: np.ndarray
+    y_test: np.ndarray
+    X_test: np.ndarray
+
+
+def test_data(resolution):
+    """
+    return some random numbers in the right structure to test the pipeline with
+    """
+    return np.random.random([11, resolution[0], resolution[1], 3]), np.array([1, 1, 1, 2, 2, 2, 3, 3, 4, 4, 5])
+
+
+def enfsi_data(resolution, year) -> Tuple[np.ndarray, np.ndarray]:
+    folder = os.path.join('resources', 'enfsi', str(year))
+    X = []
+    y = []
+    i_cls = 1 # start on 1 as ENFSI does it
+    df = pd.read_csv(os.path.join(folder, 'truth.csv')).set_index('id')
+    for file in os.listdir(folder):
+        if not file.endswith('csv'):
+            # TODO check RGB/GRB ordering/colous usage on models.
+            img = cv2.imread(os.path.join(folder, file), cv2.COLOR_BGR2RGB)
+            if year == 2011:
+                cls = int(file[:3])
+            elif year == 2012:
+                cls = int(file[:2])
+            elif year == 2013:
+                cls = int(file[1:3])
+            elif year == 2017:
+                cls = int(file[1:3])
+            else:
+                raise ValueError(f'Unknown ENFSI year {year}')
+            if df.loc[cls]['same'] == 1:
+                # same source
+                y.append(cls)
+            else:
+                #different source, make new class int
+                y.append(len(df)+i_cls)
+                i_cls+=1
+            X.append(img)
+    return np.array(X), np.array(y)
+
+
+def combine_data(dataset_callables, resolution) -> Tuple[np.ndarray, np.ndarray]:
+    """
+    gets the X and y for all data in the callables, and returns the total set
+    """
+    X = np.array([])
+    y = np.array([])
+    max_class = 0
+    for dataset_callable in dataset_callables:
+        this_X, this_y = dataset_callable(resolution)
+        y = np.append(y, this_y+max_class)
+        max_class = max(y)
+        X = np.append(X, this_X)
+    return X, y.astype(int)
+
+def get_data(dataset_callable, resolution=(100, 100),
+             train_calibration_same_data=True, fraction_calibration=None, fraction_test=0.2) -> Images:
+    """
+    Takes a function that returns X, y, with X images and y identities. Returns a dataset with all data
+    split into the right datasets
+
+
+    fraction_calibration is the fraction of train data that is used for calibration,
+        so fraction_calibration + fraction_test can be > 1
+
+    """
+
+    X, y = dataset_callable(resolution=resolution)
+    # TODO for now we will let the model resize, in future we should enforce the right resolution to come from preprocessing
```

Comment:
Willen we resize ook al in het preprocessing doen? Want ieder model heeft een ander formaat, dus dan moet er een dataset per model komen.

Reply:
ik denk het wel